src/JZSolver.cpp

```diff
@@ -0,0 +1,106 @@
+#include "JZSolver.hpp"
+#include <numeric>
+#include <iostream>
+
+bool JZSolver::solve(std::shared_ptr<Board> b)
+{
+    verifyBoardClues(b);
+    findAndFillCompleteLines(b);
+
+    while(not b->isSolved())
+    {
+        stepCounter_++;
+        // methods to solve puzzle
+        // if method changes something at board then start loop from beginning
+        // if all methods haven't changed anything then break loop and admit failure
+        break;
+    }
+    return b->isSolved();
+}
+
+unsigned int JZSolver::getStepCounter()
+{
+    return stepCounter_;
+}
+
+void JZSolver::verifyBoardClues(std::shared_ptr<Board> b)
+{
+    if(b->getCluesCols().size() != b->getSizeCols())
+        throw std::length_error("Invalid clues: some columns are missing clues.");
+
+    if(b->getCluesRows().size() != b->getSizeRows())
+        throw std::length_error("Invalid clues: some rows are missing clues.");
+
+    unsigned int sumRows = 0;
+    unsigned int sumCols = 0;
+
+    for(auto row : b->getCluesRows())
+    {
+        sumRows += std::accumulate(row.begin(), row.end(), 0);
+    }
+
+    for(auto col : b->getCluesCols())
+    {
+        sumCols += std::accumulate(col.begin(), col.end(), 0);
+    }
+
+    if(sumRows != sumCols)
+    {
+        std::string msg = "Invalid clues: different object count in rows (which is " +
+                          std::to_string(sumRows) + ") and columns (which is " +
+                          std::to_string(sumCols) + ").";
+        throw std::logic_error(msg);
+    }
+}
+
+void JZSolver::findAndFillCompleteLines(std::shared_ptr<Board> b)
+{
+#if DEBUG
```

Comment:
You should strip out this debugging mechanism before merging :)

Reply:
Even if it design specially to fullfill requirements of point 6 in project's extra functionalities?

Reply:
Yes. Visualisation should not be done on macros. You can use some design patterns to enable/disable without modifying the code of the algorithms. Template Method (also known as Non Virtual Interface) may be one of them, but not the only one. You can also use Decorator.

Reply:
So removing #define and adding bool value to class private members that would determine if visualization code is executed or not is not a good idea?

Reply:
No :) It will work of course. But you break Single Responsibility Principle. Another class or function should be responsible for visualization :)
If you just want to deliver this solution, I would not argue. However if you want to continue work on it it would be better to have it extracted. It should be done as a separate task.

Reply:
I have cut this code out. I will add step visualization in the end.